### DISCUSSION TITLE
Trying to recreate SGD network from frontiers bindsnet article

### DISCUSSION THREAD

**Discussion by dennismoolenaarwork** [BindsNET/bindsnet#524] [General]
I am trying to recreate the network in Figure 5 of the 2018 Frontiers article with title: BindsNET: A Machine Learning-Oriented Spiking Neural Networks Library in Python. ( doi: 10.3389/fninf.2018.00089)
I had to make a number of changes to allow it to run. The code is below. It runs but it is not learning. The accuracy for each step in each epoch is the exactly the same even though the weight is updated.

Changes:
- Had to change RealInput to Input. Is this correct ?
- Had to replace network.reset_ to netwok.reset_state_variables.
- Had to add [0] to summed_inputs to avoid pytorch errors
- Had to modified FashionMNIST call

I have the following questions:
- Why is the model not learning even though grads("dl/dw") is non-zero ?
- criterion is set but never used ?

According to paper this should be able to hit accuracy of 85%. Well this only hits 10%.
Does anyone know what is wrong here ?

Model:
====================================================================================
import torch
from bindsnet.network import Network
from bindsnet.datasets import FashionMNIST
from bindsnet.network.monitors import Monitor
from bindsnet.network.topology import Connection
from bindsnet.network.nodes import Input, IFNodes

# Network building.
network = Network()
input_layer = Input(n=784, sum_input=True)
output_layer = IFNodes(n=10, sum_input=True)
network.add_layer(input_layer, name='X')
network.add_layer(output_layer, name='Y')
input_connection = Connection(input_layer, output_layer, norm=150, wmin=-1, wmax=1)
network.add_connection(input_connection, source="X", target='Y')
# Stats variable monitoring,
time = 25 # No. of simulation time steps per example.
for l in network.layers:
    m = Monitor(network.layers[l], state_vars=['s'], time=time)
    network.add_monitor(m, name=l)

# Load Fashion-MNIST data.
ds = FashionMNIST(root='.', download=True, train=True)
images = ds.train_data
labels = ds.train_labels

# Run training
grads = {}
lr, lr_decay = 1e-2, 0.95

criterion = torch.nn.CrossEntropyLoss()
spike_ims, spike_axes, weights_im = None, None, None

for epoch in range(4):
    correct = 0
    for i, (image, label) in enumerate(zip(images.view(-1, 784) / 255, labels)):
        # fun simulation for stngle datus
        inpts = {'X': image.repeat(time, 1), 'Y_b': torch.ones(time, 1)}
        network.run(inputs=inpts, time=time)
        # Retrieve spikes and summed inputs from both layers.
        label = torch.tensor(label).long()
        spikes = {l: network.monitors[l].get('s') for l in network.layers}
        summed_inputs = {l: network.layers[l].summed for l in network. layers}
        # Compute softmax of output activity, get predicted label,
        output = spikes['Y'].sum(-1).float().softmax(0).view(1, -1)
        predicted = output.argmax(1).item()
        # Compute gradient of loss and do SGD update.
        grads['dl/df'] = summed_inputs['Y'].softmax(0)[0]
        grads['dl/df'][label] -= 1
        grads['dl/dw'] = torch.ger(summed_inputs['X'][0], grads['dl/df'])
        network.connections ['X', 'Y'].w -= lr * grads['dl/dw']

        if predicted == label:
            correct += 1
        # Decay learning rate.
        if i > 0 and i % 500 == 0:
            lr *= lr_decay
        if i > 0 and i % 1000 == 0:
            print(f'Epoch {epoch} Step {i} Accuracy: {correct/i} ')
        network.reset_state_variables()

**Comment by SimonInParis** [comment]
**RealInput** is not available anymore. According to the 2018 paper, _"for simulating constant current injection with the RealInput object"_.
Meaning that the real-valued dataset provides a current injection (a constant pressure to the neuron), eventually making it spike.

You can still replicate this by constantly adding the voltage to the input layer neurons by using the `injects_v` keyword in the model `run()` function, so that they will fire at a rate corresponding to the pixel brightness.
You have to use IF or LIF neurons for the input layer as well.

I think this code snippet you're referring to was a only a POC, nothing really useful.

Ideally, if you're looking for gradient descent in a SNN design with BindsNet, you will probably have to compute you own continuous differentiable function to simulate the spikes occurring in each layer. For example, see: [https://arxiv.org/pdf/1901.09948.pdf](url)

**Comment by dennismoolenaarwork** [comment]
Hi Simon,

Thank you for your response. I am not necessarily interested in SGD but in a complete example of supervised training in BindsNet.
I will keep trying to build a simple SNN example in BindsNet.


Thanx

Dennis
